core/layout_api.php

```diff
@@ -793,34 +833,64 @@ function layout_print_sidebar( $p_active_sidebar_page = null ) {
 						$t_link = 'manage_proj_page.php';
 					}
 				}
-				layout_sidebar_menu( $t_link , 'manage_link', 'fa-gears' );
+
+				$t_sidebar_items[] = array(
+					'url' => $t_link,
+					'title' => 'manage_link',
+					'icon' => 'fa-gears',
+				);
 			}
 		}
 
 		# Time Tracking / Billing
 		if( config_get( 'time_tracking_enabled' ) && access_has_project_level( config_get( 'time_tracking_reporting_threshold', $t_current_project ) ) ) {
-			layout_sidebar_menu( 'billing_page.php', 'time_tracking_billing_link', 'fa-clock-o', $p_active_sidebar_page );
+			$t_sidebar_items[] = array(
+				'url' => 'billing_page.php',
+				'title' => 'time_tracking_billing_link',
+				'icon' => 'fa-clock-o',
+			);
 		}
 
 		# Plugin / Event added options
-		$t_event_menu_options = event_signal( 'EVENT_MENU_MAIN' );
-		layout_plugin_menu_options_for_sidebar( $t_event_menu_options, $p_active_sidebar_page );
+		$t_event_menu_main = event_signal( 'EVENT_MENU_MAIN' );
+		$t_plugin_menu_items_back = layout_plugin_menu_options_for_sidebar( $t_event_menu_main );
+
+		if( is_array( $t_plugin_menu_items_back ) ) {
+			$t_sidebar_items = array_merge( $t_sidebar_items, $t_plugin_menu_items_back );
+		}
 
 		# Config based custom options
-		layout_config_menu_options_for_sidebar( $p_active_sidebar_page );
+		$t_config_menu_items = layout_config_menu_options_for_sidebar();
+
+		if( is_array( $t_config_menu_items ) ) {
+			$t_sidebar_items = array_merge( $t_sidebar_items, $t_config_menu_items );
+		}
 
-		# Ending sidebar markup
-		layout_sidebar_end();
+		# Allow plugins to alter the sidebar items array
+		$t_modified_sidebar_items = event_signal( 'EVENT_MENU_MAIN_FILTER', array( $t_sidebar_items ) );
+		if( is_array( $t_modified_sidebar_items ) && count( $t_modified_sidebar_items ) > 0 ) {
+			$t_sidebar_items = $t_modified_sidebar_items[0];
+		}
```

Comment:
As a side note, I initially found it strange that the $t_sidebar_items was wrapped in an array in the event_signal() call, and then extracted back (line 872), until I realized that it's required due to the plugin API relying on call_user_func_array().

This also means that a plugin using this event needs to do the same as well when returning the updated menu  from the hook function, i.e. `return array( $p_menu_data );` and not `return $p_menu_data;`, as I found out while testing the new event.

@JamesAtChamSys I suggest to clarify this in the documentation.

Reply:
I've done this myself.